pkg/gadget-collection/gadgets/top/ebpf/gadget.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2019-2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpf
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cilium/ebpf"
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/apis/gadget/v1alpha1"
+	"github.com/kinvolk/inspektor-gadget/pkg/bpfstats"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadget-collection/gadgets"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/ebpftop/piditer"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/ebpftop/types"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+)
+
+type Trace struct {
+	resolver gadgets.Resolver
+
+	started  bool
+	donechan chan bool
+}
+
+type TraceFactory struct {
+	gadgets.BaseFactory
+}
+
+func NewFactory() gadgets.TraceFactory {
+	return &TraceFactory{
+		BaseFactory: gadgets.BaseFactory{DeleteTrace: deleteTrace},
+	}
+}
+
+func (f *TraceFactory) Description() string {
+	t := `ebpftop shows cpu time used by ebpf programs.
+
+The following parameters are supported:
+ - %s: Output interval, in seconds. (default %d)
+ - %s: Maximum rows to print. (default %d)
+ - %s: The field to sort the results by (%s). (default %s)`
+	return fmt.Sprintf(t, types.IntervalParam, types.IntervalDefault,
+		types.MaxRowsParam, types.MaxRowsDefault,
+		types.SortByParam, strings.Join(types.SortBySlice, ","), types.SortByDefault)
```

Comment:
We have not detailed parameters like this in other gadgets. This requires `types.MaxRowsDefault` to be defined in  `pkg/gadgets/ebpftop/types` even though it's specific to the kubectl-gadget CLI, so it makes me uneasy...

Also this Description() function generates the documentation for interacting with the Trace CR directly, it's not about the documentation for how to use the `kubectl-gadget` CLI (which would be in `docs/guides/top/` instead).

Arguably, separating the documentation for the different Trace CR and for the different `kubectl-gadget` commands is not ideal, but I think it should be addressed for all gadgets in a separate PR.

cc @blanquicet

Reply:
> We have not detailed parameters like this in other gadgets. 
>
All the gadget in the top category has the same detailed parameters, e.g., in in [file gadget](https://github.com/kinvolk/inspektor-gadget/blob/04c2a3664d830cfc39c294f9a37b1b73159de008/pkg/gadget-collection/gadgets/top/file/gadget.go#L53-L57). I think it is useful for third-party applications that will explore the documentation on [docs/gadgets](https://github.com/kinvolk/inspektor-gadget/tree/main/docs/gadgets), don't you think?

However, I agree that we it is not idea to have it in two places. Maybe we could move it to a common place where it keeps sync between kubectl-gadget and the CRD description? Of course, in another PR.

> This requires types.MaxRowsDefault to be defined in pkg/gadgets/ebpftop/types even though it's specific to the kubectl-gadget CLI
>
It shouldn't be specific for kubectl-gadget because `types.IntervalParam`, `types.MaxRowsParam` and  `types.SortByParam` should be used to extract the params, e.g. in [file gadget](https://github.com/kinvolk/inspektor-gadget/blob/e8250746a361a67bf90cdb2dee48cfe730693bba/pkg/gadget-collection/gadgets/top/file/gadget.go#L113-L148).

Reply:
Oh, I didn't realise those parameters were in the Trace custom resource. Ok then, you can ignore my comment.